service/controller/v3/resource/instance/create.go

```diff
@@ -248,58 +246,141 @@ func (r *Resource) allInstances(ctx context.Context, customObject providerv1alph
 	var instanceToReimage *compute.VirtualMachineScaleSetVM
 	{
-		r.logger.LogCtx(ctx, "level", "debug", "message", "looking for the next instance to be updated or reimaged")
+		r.logger.LogCtx(ctx, "level", "debug", "message", "looking for the next instance to be updated, drained or reimaged")
 
-		instanceToUpdate, instanceToReimage, err = findActionableInstance(customObject, instances, versionValue)
+		instanceToUpdate, instanceToDrain, instanceToReimage, err = findActionableInstance(customObject, instances, nodeConfigs, instanceNameFunc, versionValue)
 		if IsVersionBlobEmpty(err) {
 			// When no version bundle version is found it means the cluster just got
 			// created and the version bundle versions are not yet tracked within the
 			// parameters of the guest cluster's VMSS deployment. In this case we must
 			// not select an instance to be reimaged because we would roll a node that
 			// just got created and is already up to date.
-			r.logger.LogCtx(ctx, "level", "debug", "message", "version blob still empty")
-			return nil, nil, nil
+			r.logger.LogCtx(ctx, "level", "debug", "message", "no instance found to be updated, drained or reimaged")
+			return nil, nil, nil, nil
 		} else if err != nil {
-			return nil, nil, microerror.Mask(err)
+			return nil, nil, nil, microerror.Mask(err)
 		}
 
-		if instanceToUpdate == nil && instanceToReimage == nil {
+		if instanceToUpdate == nil && instanceToDrain == nil && instanceToReimage == nil {
 			// Neither did we find an instance to be updated nor to be reimaged.
 			// Nothing has to be done or we already processes all instances.
-			r.logger.LogCtx(ctx, "level", "debug", "message", "no instance found to be updated or reimaged")
-			return nil, nil, nil
+			r.logger.LogCtx(ctx, "level", "debug", "message", "no instance found to be updated, drained or reimaged")
+			return nil, nil, nil, nil
 		}
 
 		if instanceToUpdate != nil {
 			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found instance '%s' has to be updated", instanceNameFunc(customObject, *instanceToUpdate.InstanceID)))
 		}
+		if instanceToDrain != nil {
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found instance '%s' has to be drained", instanceNameFunc(customObject, *instanceToDrain.InstanceID)))
+		}
 		if instanceToReimage != nil {
 			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found instance '%s' has to be reimaged", instanceNameFunc(customObject, *instanceToReimage.InstanceID)))
 		}
 	}
 
-	return instanceToUpdate, instanceToReimage, nil
+	return instanceToUpdate, instanceToDrain, instanceToReimage, nil
 }
 
 func (r *Resource) reimageInstance(ctx context.Context, customObject providerv1alpha1.AzureConfig, instance *compute.VirtualMachineScaleSetVM, deploymentNameFunc func(customObject providerv1alpha1.AzureConfig) string, instanceNameFunc func(customObject providerv1alpha1.AzureConfig, instanceID string) string) error {
```

Comment:
Unrelated: Isn't it just simpler to pass `deploymentName string` instead of `deploymentNameFunc func(customObject providerv1alpha1.AzureConfig) string`?

Reply:
Maybe. I decided against it because the calling side would be more bloated and ugly.